compiler/optimizer/abstractinterpreter/AbsValue.hpp

```diff
@@ -0,0 +1,160 @@
+    * @return AbsValue*
+    * 
+    * The return value is self or NULL.
+    * When succeeding to merge, it returns self.
+    * NULL denotes the merge fails (when merging with different dataTypes or merging with a Dummy AbsValue).
+    */
+   AbsValue* merge(AbsValue *other, OMR::ValuePropagation *vp);
+
+   /**
+    * @brief Check whether the AbsValue has no constraint
+    *
+    * @return bool
+    */
+   bool isTop() { return _constraint == NULL; }
+
+   /**
+    * @brief Set the constriant to unknown.
+    *
+    * @return void
+    */
+   void setToTop() { _constraint = NULL; }
+
+   /**
+    * @brief Check if the AbsValue is only used to take a slot as the second word of the 64-bit data types
+    *
+    * @return bool
+    */
+   bool isDummy() { return _isDummy; }
+
+   /**
+    * @brief Check if the AbsValue is a parameter.
+    *
+    * @return bool
+    */
+   bool isParameter() { return _paramPos >= 0; }
+
+   /**
+    * @brief Check if the AbsValue is 'this' (the implicit parameter.)
+    *
+    * @return bool
+    */   
+   bool isImplicitParameter() { return _paramPos == 0 && _isImplicitParameter; }
+
+   /**
+    * A set of methods for creating different kinds of AbsValue.
+    * OMR::ValuePropagation is required as a parameter since we are using TR::VPConstraint as the constraint.
+    * Though it has nothing to do with the Value Propagation optimization.
+    */
+   static AbsValue* create(TR::VPConstraint *constraint, TR::DataType dataType, TR::Region& region);
+
+   static AbsValue* create(AbsValue* other, TR::Region& region);
+   
+   static AbsValue* createClassObject(TR_OpaqueClassBlock* opaqueClass, bool mustBeNonNull, TR::Region& region, OMR::ValuePropagation* vp);
+
+   static AbsValue* createNullObject(TR::Region& region, OMR::ValuePropagation* vp);
+   static AbsValue* createArrayObject(TR_OpaqueClassBlock* arrayClass, bool mustBeNonNull, int32_t lengthLow, int32_t lengthHigh, int32_t elementSize, TR::Region& region, OMR::ValuePropagation* vp);
+   
+   static AbsValue* createStringConst(TR::SymbolReference* symRef, TR::Region& region, OMR::ValuePropagation* vp);
+
+   static AbsValue* createIntConst(int32_t value, TR::Region& region, OMR::ValuePropagation* vp);
+   static AbsValue* createLongConst(int64_t value, TR::Region& region, OMR::ValuePropagation* vp);
+   static AbsValue* createIntRange(int32_t low, int32_t high, TR::Region& region, OMR::ValuePropagation* vp);
+   static AbsValue* createLongRange(int64_t low, int64_t high, TR::Region& region, OMR::ValuePropagation* vp);
+
+   static AbsValue* createTopInt(TR::Region& region);
+   static AbsValue* createTopLong(TR::Region& region);
+
+   static AbsValue* createTopObject(TR::Region& region);
+
+   static AbsValue* createTopDouble(TR::Region& region);
+   static AbsValue* createTopFloat(TR::Region& region);
+
+   static AbsValue* createDummyDouble(TR::Region& region);
+   static AbsValue* createDummyLong(TR::Region& region);
```

Comment:
All of these APIs seem to have `TR::Region& region` as a common parameter. To me it would make sense for this to be the first parameter for all of these APIs so as to introduce some consistency. For the APIs using `OMR::ValuePropagation* vp`, that parameter should likely come next.